File: trade_dashboard/app/api/trading.py
"""
트레이딩 관련 API 엔드포인트
"""
from fastapi import APIRouter, HTTPException, Depends, Query
from typing import List, Dict, Any, Optional
from datetime import datetime, date
from pydantic import BaseModel, Field
from enum import Enum

router = APIRouter()


class TradingMode(str, Enum):
    """트레이딩 모드"""
    LIVE = "live"
    PAPER = "paper"
    BACKTEST = "backtest"


class TradingStatus(str, Enum):
    """트레이딩 상태"""
    IDLE = "idle"
    RUNNING = "running"
    PAUSED = "paused"
    STOPPED = "stopped"
    ERROR = "error"


class OrderType(str, Enum):
    """주문 유형"""
    MARKET = "MKT"
    LIMIT = "LMT"
    STOP = "STP"
    STOP_LIMIT = "STP_LMT"


class OrderSide(str, Enum):
    """주문 방향"""
    BUY = "BUY"
    SELL = "SELL"


class OrderStatus(str, Enum):
    """주문 상태"""
    PENDING = "pending"
    SUBMITTED = "submitted"
    FILLED = "filled"
    CANCELLED = "cancelled"
    REJECTED = "rejected"


class TradingSystemStatus(BaseModel):
    """트레이딩 시스템 상태"""
    mode: TradingMode
    status: TradingStatus
    active_strategies: List[str]
    ibkr_connected: bool
    last_heartbeat: datetime
    uptime_seconds: int
    total_positions: int
    total_orders_today: int


class OrderRequest(BaseModel):
    """주문 요청 모델"""
    symbol: str
    side: OrderSide
    quantity: int = Field(gt=0)
    order_type: OrderType = OrderType.MARKET
    limit_price: Optional[float] = None
    stop_price: Optional[float] = None
    time_in_force: str = "GTC"
    strategy_name: Optional[str] = None


class OrderResponse(BaseModel):
    """주문 응답 모델"""
    order_id: str
    symbol: str
    side: OrderSide
    quantity: int
    order_type: OrderType
    status: OrderStatus
    submitted_at: datetime
    filled_at: Optional[datetime] = None
    avg_fill_price: Optional[float] = None
    commission: Optional[float] = None


class AccountInfo(BaseModel):
    """계좌 정보 모델"""
    account_id: str
    total_cash: float
    net_liquidation: float
    buying_power: float
    total_positions_value: float
    total_pnl: float
    daily_pnl: float
    margin_used: float
    margin_available: float
    currency: str = "USD"
    last_updated: datetime


@router.get("/status", response_model=TradingSystemStatus)
async def get_trading_status():
    """트레이딩 시스템 상태 조회"""
    # TODO: 실제 시스템 상태 조회
    return {
        "mode": TradingMode.PAPER,
        "status": TradingStatus.RUNNING,
        "active_strategies": ["Example1Strategy"],
        "ibkr_connected": True,
        "last_heartbeat": datetime.now(),
        "uptime_seconds": 3600,
        "total_positions": 5,
        "total_orders_today": 12
    }


@router.post("/start")
async def start_trading(mode: TradingMode = TradingMode.PAPER):
    """트레이딩 시작"""
    # TODO: trade_engine에 트레이딩 시작 요청
    return {
        "message": f"Trading started in {mode} mode",
        "mode": mode,
        "started_at": datetime.now()
    }


@router.post("/stop")
async def stop_trading():
    """트레이딩 중지"""
    # TODO: trade_engine에 트레이딩 중지 요청
    return {
        "message": "Trading stopped",
        "stopped_at": datetime.now()
    }


@router.post("/pause")
async def pause_trading():
    """트레이딩 일시정지"""
    # TODO: trade_engine에 일시정지 요청
    return {
        "message": "Trading paused",
        "paused_at": datetime.now()
    }


@router.post("/resume")
async def resume_trading():
    """트레이딩 재개"""
    # TODO: trade_engine에 재개 요청
    return {
        "message": "Trading resumed",
        "resumed_at": datetime.now()
    }


@router.get("/account", response_model=AccountInfo)
async def get_account_info():
    """계좌 정보 조회"""
    # TODO: IBKR에서 실제 계좌 정보 조회
    return {
        "account_id": "DU1234567",
        "total_cash": 100000.0,
        "net_liquidation": 150000.0,
        "buying_power": 200000.0,
        "total_positions_value": 50000.0,
        "total_pnl": 5000.0,
        "daily_pnl": 500.0,
        "margin_used": 25000.0,
        "margin_available": 75000.0,
        "currency": "USD",
        "last_updated": datetime.now()
    }


@router.post("/orders", response_model=OrderResponse)
async def place_order(order: OrderRequest):
    """수동 주문 실행"""
    # TODO: IBKR에 주문 전송
    return {
        "order_id": "ORD123456",
        "symbol": order.symbol,
        "side": order.side,
        "quantity": order.quantity,
        "order_type": order.order_type,
        "status": OrderStatus.SUBMITTED,
        "submitted_at": datetime.now(),
        "filled_at": None,
        "avg_fill_price": None,
        "commission": None
    }


@router.get("/orders", response_model=List[OrderResponse])
async def get_orders(
    status: Optional[OrderStatus] = None,
    symbol: Optional[str] = None,
    start_date: Optional[date] = None,
    end_date: Optional[date] = None,
    limit: int = Query(100, le=1000)
):
    """주문 목록 조회"""
    # TODO: 실제 주문 데이터 조회
    orders = []
    return orders


@router.delete("/orders/{order_id}")
async def cancel_order(order_id: str):
    """주문 취소"""
    # TODO: IBKR에 주문 취소 요청
    return {
        "message": f"Order {order_id} cancelled",
        "cancelled_at": datetime.now()
    }


@router.post("/orders/cancel-all")
async def cancel_all_orders():
    """모든 미체결 주문 취소"""
    # TODO: 모든 미체결 주문 취소
    return {
        "message": "All pending orders cancelled",
        "cancelled_count": 0
    }


@router.post("/positions/close-all")
async def close_all_positions():
    """모든 포지션 청산"""
    # TODO: 모든 포지션 청산
    return {
        "message": "All positions closed",
        "closed_count": 0
    }


@router.get("/pnl/daily")
async def get_daily_pnl(days: int = 30):
    """일별 손익 조회"""
    # TODO: 실제 일별 손익 데이터 조회
    return {
        "days": days,
        "pnl_data": []
    }

File: trade_frontend/src/lib/api/trading.ts
import { api } from './client';

export type TradingMode = 'live' | 'paper' | 'backtest';
export type TradingStatus = 'idle' | 'running' | 'paused' | 'stopped' | 'error';
export type OrderType = 'MKT' | 'LMT' | 'STP' | 'STP_LMT';
export type OrderSide = 'BUY' | 'SELL';
export type OrderStatus = 'pending' | 'submitted' | 'filled' | 'cancelled' | 'rejected';

export interface TradingSystemStatus {
  mode: TradingMode;
  status: TradingStatus;
  active_strategies: string[];
  ibkr_connected: boolean;
  last_heartbeat: string;
  uptime_seconds: number;
  total_positions: number;
  total_orders_today: number;
}

export interface AccountInfo {
  account_id: string;
  total_cash: number;
  net_liquidation: number;
  buying_power: number;
  total_positions_value: number;
  total_pnl: number;
  daily_pnl: number;
  margin_used: number;
  margin_available: number;
  currency: string;
  last_updated: string;
}

export interface OrderRequest {
  symbol: string;
  side: OrderSide;
  quantity: number;
  order_type: OrderType;
  limit_price?: number;
  stop_price?: number;
  time_in_force?: string;
  strategy_name?: string;
}

export interface OrderResponse {
  order_id: string;
  symbol: string;
  side: OrderSide;
  quantity: number;
  order_type: OrderType;
  status: OrderStatus;
  submitted_at: string;
  filled_at?: string;
  avg_fill_price?: number;
  commission?: number;
}

export const tradingApi = {
  // 시스템 상태 조회
  getStatus: async (): Promise<TradingSystemStatus> => {
    const response = await api.get<TradingSystemStatus>('/api/trading/status');
    if (response.error) throw new Error(response.error);
    return response.data!;
  },

  // 트레이딩 시작
  startTrading: async (mode: TradingMode = 'paper'): Promise<any> => {
    const response = await api.post<any>(`/api/trading/start?mode=${mode}`, {});
    if (response.error) throw new Error(response.error);
    return response.data!;
  },

  // 트레이딩 중지
  stopTrading: async (): Promise<any> => {
    const response = await api.post<any>('/api/trading/stop', {});
    if (response.error) throw new Error(response.error);
    return response.data!;
  },

  // 트레이딩 일시정지
  pauseTrading: async (): Promise<any> => {
    const response = await api.post<any>('/api/trading/pause', {});
    if (response.error) throw new Error(response.error);
    return response.data!;
  },

  // 트레이딩 재개
  resumeTrading: async (): Promise<any> => {
    const response = await api.post<any>('/api/trading/resume', {});
    if (response.error) throw new Error(response.error);
    return response.data!;
  },

  // 계좌 정보 조회
  getAccountInfo: async (): Promise<AccountInfo> => {
    const response = await api.get<AccountInfo>('/api/trading/account');
    if (response.error) throw new Error(response.error);
    return response.data!;
  },

  // 주문 실행
  placeOrder: async (order: OrderRequest): Promise<OrderResponse> => {
    const response = await api.post<OrderResponse>('/api/trading/orders', order);
    if (response.error) throw new Error(response.error);
    return response.data!;
  },

  // 주문 목록 조회
  getOrders: async (params?: {
    status?: OrderStatus;
    symbol?: string;
    start_date?: string;
    end_date?: string;
    limit?: number;
  }): Promise<OrderResponse[]> => {
    const queryParams = new URLSearchParams();
    if (params?.status) queryParams.append('status', params.status);
    if (params?.symbol) queryParams.append('symbol', params.symbol);
    if (params?.start_date) queryParams.append('start_date', params.start_date);
    if (params?.end_date) queryParams.append('end_date', params.end_date);
    if (params?.limit) queryParams.append('limit', params.limit.toString());
    
    const response = await api.get<OrderResponse[]>(`/api/trading/orders?${queryParams}`);
    if (response.error) throw new Error(response.error);
    return response.data!;
  },

  // 주문 취소
  cancelOrder: async (orderId: string): Promise<any> => {
    const response = await api.delete<any>(`/api/trading/orders/${orderId}`);
    if (response.error) throw new Error(response.error);
    return response.data!;
  },

  // 모든 주문 취소
  cancelAllOrders: async (): Promise<any> => {
    const response = await api.post<any>('/api/trading/orders/cancel-all', {});
    if (response.error) throw new Error(response.error);
    return response.data!;
  },

  // 모든 포지션 청산
  closeAllPositions: async (): Promise<any> => {
    const response = await api.post<any>('/api/trading/positions/close-all', {});
    if (response.error) throw new Error(response.error);
    return response.data!;
  },

  // 일별 손익 조회
  getDailyPnL: async (days: number = 30): Promise<any> => {
    const response = await api.get<any>(`/api/trading/pnl/daily?days=${days}`);
    if (response.error) throw new Error(response.error);
    return response.data!;
  }
};